testing/upload_validation.ts

```diff
@@ -480,18 +480,126 @@ const booleanPropertySchema = zodCompleteObject<BooleanSchema & ObjectSchemaProp
+  type: zodDiscriminant(ValueType.Number),
+  codaType: zodDiscriminant(ValueHintType.Date),
+  format: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const numericTimePropertySchema = zodCompleteObject<NumericTimeSchema & ObjectSchemaProperty>({
   type: zodDiscriminant(ValueType.Number),
-  codaType: z.enum([...NumberHintValueTypes]).optional(),
+  codaType: zodDiscriminant(ValueHintType.Time),
+  format: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const numericDateTimePropertySchema = zodCompleteObject<NumericDateTimeSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.Number),
+  codaType: zodDiscriminant(ValueHintType.DateTime),
+  dateFormat: z.string().optional(),
+  timeFormat: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const numberPropertySchema = z.union([
+  numericPropertySchema, 
+  scalePropertySchema, 
+  sliderPropertySchema, 
+  currencyPropertySchema,
+  numericDatePropertySchema,
+  numericTimePropertySchema,
+  numericDateTimePropertySchema,
+  baseNumberPropertySchema,
+]);
+
+const numericPackFormulaSchema = zodCompleteObject<Omit<NumericPackFormula<any>, 'execute'>>({
+  ...commonPackFormulaSchema,
+  resultType: zodDiscriminant(Type.number),
+  schema: numberPropertySchema.optional(),
+});
+
+const stringDatePropertySchema = zodCompleteObject<StringDateSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.String),
+  codaType: zodDiscriminant(ValueHintType.Date),
+  format: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const stringTimePropertySchema = zodCompleteObject<StringTimeSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.String),
+  codaType: zodDiscriminant(ValueHintType.Time),
+  format: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const stringDateTimePropertySchema = zodCompleteObject<StringDateTimeSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.String),
+  codaType: zodDiscriminant(ValueHintType.DateTime),
+  dateFormat: z.string().optional(),
+  timeFormat: z.string().optional(),
+  ...basePropertyValidators,
+})
+
+const durationPropertySchema = zodCompleteObject<DurationSchema & ObjectSchemaProperty>({
+  type: zodDiscriminant(ValueType.String),
+  codaType: zodDiscriminant(ValueHintType.Duration),
+  precision: z.number().optional(),
+  maxUnit: z.nativeEnum(DurationUnit).optional(),
   ...basePropertyValidators,
 });
 
-const stringPropertySchema = zodCompleteObject<StringSchema & ObjectSchemaProperty>({
+const baseStringPropertySchema = zodCompleteObject<BaseStringSchema & ObjectSchemaProperty>({
   type: zodDiscriminant(ValueType.String),
   codaType: z.enum([...StringHintValueTypes]).optional(),
   ...basePropertyValidators,
 });
 
+const stringPropertySchema = z.union([
+  stringDatePropertySchema,
+  stringTimePropertySchema,
+  stringDateTimePropertySchema,
+  durationPropertySchema,
+  baseStringPropertySchema
```

Comment:
I think this is a little over-permissive. Because this validator allows all of the string hint types, for the date/time/duration one's it's actually redundant with the specific validators you've already declared, and allows schemas with those hint types to ALSO match this validator, which is more permissive. So I think we should probably have a fall-through validator that's more strict about which codaTypes it will validate. I made that change along with a few others here, you can just borrow from this commit (or I can commit it to your branch or you can cherrypick it)

https://github.com/coda/packs-sdk/commit/8249aa6763985634ce2c1d77d0f749bbdb7e0c3f

Reply:
Just so I understand, this is just to make the typing cleaner (and not have overlapping types)? And it shouldn't really affect any edge cases, right?

Reply:
Well, the way it was before was actually negatively impacting edge cases. The union validator had validators for things like:

```
type: ValueType.String;
codaType: ValueHintType.Date;
format?: string;
```

and

```
type: ValueType.String;
codaType?: ValueHintType;
```

Suppose that `format` became a required field for schemas with date hint types. If you pass a schema with ValueHintType.Date but no format, zod should fail. But because baseStringPropertySchema looked like the bottom schema, it would allow that invalid schema through, because while it would fail the date validator, it would match the base validator.

In other words, need to make sure the types and validators are not just properly inclusive, but also properly exclusive.